app/src/main/java/org/onionshare/android/ShareManager.kt

```diff
@@ -1,221 +1,155 @@
+            emit(ShareUiState.Error(f.files))
         } else {
-            _shareState.value = ShareUiState.FilesAdded(newList)
-        }
-    }
-
-    fun removeAll() {
-        // release persistable Uri permissions again
-        shareState.value.files.forEach { file ->
-            file.releaseUriPermission()
+            LOG.error("Unhandled state: ↑")
         }
-        _shareState.value = ShareUiState.NoFiles
-    }
+    }.distinctUntilChanged().onEach {
+        LOG.debug("New state: ${it::class.simpleName}")
+    }.stateIn(GlobalScope, SharingStarted.Lazily, ShareUiState.NoFiles)
 
     suspend fun onStateChangeRequested() = when (shareState.value) {
         is ShareUiState.FilesAdded -> startSharing()
         is ShareUiState.Starting -> stopSharing()
         is ShareUiState.Sharing -> stopSharing()
         is ShareUiState.Complete -> startSharing()
+        is ShareUiState.ErrorAddingFile -> startSharing()
         is ShareUiState.Error -> startSharing()
-        ShareUiState.NoFiles -> error("Sheet button should not be visible with no files")
+        is ShareUiState.Stopping -> error("Pressing sheet button while stopping should not be possible")
+        is ShareUiState.NoFiles -> error("Sheet button should not be visible with no files")
     }
 
-    @OptIn(DelicateCoroutinesApi::class)
     @Suppress("BlockingMethodInNonBlockingContext")
     private suspend fun startSharing() {
         if (startSharingJob?.isActive == true) {
             // TODO check if this always works as expected
             startSharingJob?.cancelAndJoin()
         }
+        shouldStop.value = false
         // Attention: We'll launch sharing in Global scope, so it survives ViewModel death
-        startSharingJob = GlobalScope.launch(Dispatchers.IO) {
-            val files = shareState.value.files
-            // call ensureActive() before any heavy work to ensure we don't continue when cancelled
-            ensureActive()
-            _shareState.value = ShareUiState.Starting(files, 0, 0)
-            try {
-                // TODO we might want to look into parallelizing what happens below (async {} ?)
-                // When the current scope gets cancelled, the async routine gets cancelled as well
-                ensureActive()
-                var sendPage: SendPage? = null
-                fileManager.zipFiles(files).collect { state ->
-                    ensureActive()
-                    _shareState.value = ShareUiState.Starting(files, state.progress, 0)
-                    if (state.complete) sendPage = getSendPage(state)
-                }
-                val page = sendPage ?: error("SendPage was null")
+        startSharingJob = coroutineScope {
+            launch(Dispatchers.IO) {
+                // call ensureActive() before any heavy work to ensure we don't continue when cancelled
                 ensureActive()
+                // When the current scope gets cancelled, the async routine gets cancelled as well
```

Comment:
Does the same apply to torManager.start() below? What are the implications of cancelling that call?

Reply:
Yes, there a `JobCancellationException` gets thrown which we catch and call `stop()` which stops Tor again.